File: default/server/pub/_scripts/gfx/Engine.js
(function() {
  var smio;
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };
  smio = global.smoothio;
  smio.gfx.Engine = (function() {
    function Engine(ctl, cid) {
      this.ctl = ctl;
      this.updateCanvasSize = __bind(this.updateCanvasSize, this);
      this.pushMatrix = __bind(this.pushMatrix, this);
      this.popMatrix = __bind(this.popMatrix, this);
      this.play = __bind(this.play, this);
      this.handleMouseMove = __bind(this.handleMouseMove, this);
      this.handleKeyUp = __bind(this.handleKeyUp, this);
      this.handleKeyDown = __bind(this.handleKeyDown, this);
      this.isKeyPressed = __bind(this.isKeyPressed, this);
      this.getSightDistance = __bind(this.getSightDistance, this);
      this.initEngineShaders = __bind(this.initEngineShaders, this);
      this.initEngineBuffers = __bind(this.initEngineBuffers, this);
      this.initEngine = __bind(this.initEngine, this);
      this.drawMesh = __bind(this.drawMesh, this);
      this.draw = __bind(this.draw, this);
      this.createShader = __bind(this.createShader, this);
      this.createVertexShader = __bind(this.createVertexShader, this);
      this.createFragmentShader = __bind(this.createFragmentShader, this);
      this.drawTimes = [];
      this.lastDrawTime = 0;
      this.pressedKeys = [];
      this.matrixStack = [];
      this.shaders = {};
      this.meshes = [new smio.gfx.MeshCube(this)];
      if ((this.canvas = $("#" + cid)) && this.canvas.length && (this.canvEl = this.canvas[0]) && this.initEngine() && this.requestAnimFrame) {
        this.texMan = new smio.gfx.TextureManager(this);
        this.texMan.load('stones', '/_/file/images/textures/stones.jpg');
        this.texMan.load('wood', '/_/file/images/textures/wood.jpg');
        this.texMan.load('sky3', '/_/file/images/textures/skydn.jpg');
        this.updateCanvasSize();
        this.play();
        return;
        this.addScene(this.scene = new CL3D.Scene());
        this.scene.setBackgroundColor(CL3D.createColor(255, 0, 0, 48));
        this.scene.getRootSceneNode().addChild(this.skyBox = new CL3D.SkyBoxSceneNode());
        this.skyBox.getMaterial(5).Tex1 = this.getTextureManager().getTexture('/_/file/images/textures/stars.jpg', true);
        this.skyBox.getMaterial(4).Tex1 = this.getTextureManager().getTexture('/_/file/images/textures/skxup.jpg', true);
        this.skyBox.getMaterial(0).Tex1 = this.getTextureManager().getTexture('/_/file/images/textures/skx1.jpg', true);
        this.skyBox.getMaterial(2).Tex1 = this.getTextureManager().getTexture('/_/file/images/textures/skx3.jpg', true);
        this.skyBox.getMaterial(1).Tex1 = this.getTextureManager().getTexture('/_/file/images/textures/skx2.jpg', true);
        this.skyBox.getMaterial(3).Tex1 = this.getTextureManager().getTexture('/_/file/images/textures/skx0.jpg', true);
        this.scene.getRootSceneNode().addChild(this.universe = new smio.gfx.UniverseSceneNode(this));
        this.scene.setActiveCamera(this.universe.cam);
        this.universe.camSettings(this.canvEl.width / this.canvEl.height, CL3D.degToRad(45), this.getSightDistance(20), 1);
      }
    }
    Engine.prototype.createFragmentShader = function(src) {
      return this.createShader(this.gl.FRAGMENT_SHADER, src);
    };
    Engine.prototype.createVertexShader = function(src) {
      return this.createShader(this.gl.VERTEX_SHADER, src);
    };
    Engine.prototype.createShader = function(type, src) {
      var gl, shader;
      if (gl = this.gl) {
        shader = gl.createShader(type);
        gl.shaderSource(shader, src);
        gl.compileShader(shader);
        if (!(gl.getShaderParameter(shader, gl.COMPILE_STATUS))) {
          alert(gl.getShaderInfoLog(shader));
        }
      }
      return shader;
    };
    Engine.prototype.createVertex = function(x, y, z, s, t) {
      var v;
      v = new CL3D.Vertex3D(true);
      v.Pos.X = x;
      v.Pos.Y = y;
      v.Pos.Z = z;
      v.TCoords.X = s;
      v.TCoords.Y = t;
      return v;
    };
    Engine.prototype.draw = function(timings) {
      var canvas, gl, mesh, name, shaderProg, _i, _len, _ref, _ref2, _results;
      if ((gl = this.gl) && (canvas = gl.canvas)) {
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
        mat4.identity(this.modelViewMatrix);
        _ref = this.shaders;
        for (name in _ref) {
          shaderProg = _ref[name];
          gl.uniformMatrix4fv(shaderProg.uniforms.pMatrix, false, this.projectionMatrix);
        }
        _ref2 = this.meshes;
        _results = [];
        for (_i = 0, _len = _ref2.length; _i < _len; _i++) {
          mesh = _ref2[_i];
          _results.push(!mesh.hidden ? this.drawMesh(gl, mesh, timings) : void 0);
        }
        return _results;
      }
    };
    Engine.prototype.drawMesh = function(gl, mesh, timings) {
      var name, shader, shaderProg, _ref, _ref2, _ref3, _ref4, _ref5;
      this.pushMatrix();
      mesh.beforeDraw(gl, timings);
      mat4.translate(this.modelViewMatrix, [mesh.posX, mesh.posY, mesh.posZ]);
      if (mesh.rotX) {
        mat4.rotateX(this.modelViewMatrix, mesh.rotX);
      }
      if (mesh.rotY) {
        mat4.rotateY(this.modelViewMatrix, mesh.rotY);
      }
      if (mesh.rotZ) {
        mat4.rotateZ(this.modelViewMatrix, mesh.rotZ);
      }
      if (mesh.vertexBuffer) {
        gl.bindBuffer(gl.ARRAY_BUFFER, mesh.vertexBuffer);
        _ref = this.shaders;
        for (name in _ref) {
          shader = _ref[name];
          gl.vertexAttribPointer(shader.atts.aVertexPosition, mesh.vertices[0].length, gl.FLOAT, false, 0, 0);
        }
      }
      if (mesh.colorBuffer) {
        gl.bindBuffer(gl.ARRAY_BUFFER, mesh.colorBuffer);
        _ref2 = this.shaders;
        for (name in _ref2) {
          shader = _ref2[name];
          gl.vertexAttribPointer(shader.atts.aVertexColor, mesh.colors[0].length, gl.FLOAT, false, 0, 0);
        }
      }
      if (mesh.texCoordsBuffer) {
        gl.bindBuffer(gl.ARRAY_BUFFER, mesh.texCoordsBuffer);
        _ref3 = this.shaders;
        for (name in _ref3) {
          shader = _ref3[name];
          gl.vertexAttribPointer(shader.atts.aTexCoord, mesh.texCoords[0].length, gl.FLOAT, false, 0, 0);
        }
        gl.activeTexture(gl.TEXTURE0);
        gl.bindTexture(gl.TEXTURE_2D, this.texMan.textures['sky3']);
        _ref4 = this.shaders;
        for (name in _ref4) {
          shader = _ref4[name];
          gl.uniform1i(shader.uniforms.uSampler, 0);
        }
      }
      if (mesh.indexBuffer) {
        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, mesh.indexBuffer);
      }
      _ref5 = this.shaders;
      for (name in _ref5) {
        shaderProg = _ref5[name];
        gl.uniformMatrix4fv(shaderProg.uniforms.mvMatrix, false, this.modelViewMatrix);
      }
      mesh.draw(gl, timings);
      return this.popMatrix();
    };
    Engine.prototype.initEngine = function() {
      var canvas, gl, name, names, _i, _len, _ref;
      _ref = [this.canvEl, null, ['webgl', 'experimental-webgl', 'webkit-3d', 'moz-webgl']], canvas = _ref[0], gl = _ref[1], names = _ref[2];
      this.requestAnimFrame = window.requestAnimationFrame || window.webkitRequestAnimationFrame || window.mozRequestAnimationFrame || window.oRequestAnimationFrame || window.msRequestAnimationFrame;
      this.modelViewMatrix = mat4.create();
      this.projectionMatrix = mat4.create();
      for (_i = 0, _len = names.length; _i < _len; _i++) {
        name = names[_i];
        try {
          gl = canvas.getContext(name, {
            alpha: false,
            depth: true,
            stencil: true,
            antialias: true,
            premultipliedAlpha: false,
            preserveDrawingBuffer: false
          });
        } catch (_e) {}
        if (gl) {
          break;
        }
      }
      if ((this.gl = gl)) {
        this.initEngineShaders();
        this.initEngineBuffers();
        gl.clearColor(0.1, 0.2, 0.3, 1.0);
        gl.enable(gl.DEPTH_TEST);
      }
      return gl;
    };
    Engine.prototype.initEngineBuffers = function() {
      var gl, mesh, _i, _len, _ref, _results;
      if ((gl = this.gl)) {
        _ref = this.meshes;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          mesh = _ref[_i];
          _results.push(mesh.updateBuffers());
        }
        return _results;
      }
    };
    Engine.prototype.initEngineShaders = function() {
      var attName, fragShader, gl, name, prog, shader, uniName, vertexShader, _ref, _results;
      if ((gl = this.gl)) {
        _ref = smio.gfx.Shaders;
        _results = [];
        for (name in _ref) {
          shader = _ref[name];
          _results.push((function() {
            var _i, _j, _len, _len2, _ref2, _ref3, _results2;
            if (!shader.disabled) {
              vertexShader = this.createVertexShader(shader.vertex);
              fragShader = this.createFragmentShader(shader.fragment);
              this.shaders[name] = prog = gl.createProgram();
              gl.attachShader(prog, vertexShader);
              gl.attachShader(prog, fragShader);
              gl.linkProgram(prog);
              if (!gl.getProgramParameter(prog, gl.LINK_STATUS)) {
                alert('Could not link shader program.');
              }
              gl.useProgram(prog);
              if ((!shader.atts) || !shader.atts.length) {
                shader.atts = ['aVertexPosition'];
              } else if (!_.contains(shader.atts, 'aVertexPosition')) {
                shader.atts.push('aVertexPosition');
              }
              prog.atts = {};
              prog.uniforms = {
                pMatrix: gl.getUniformLocation(prog, 'uPMatrix'),
                mvMatrix: gl.getUniformLocation(prog, 'uMVMatrix')
              };
              _ref2 = shader.uniforms;
              for (_i = 0, _len = _ref2.length; _i < _len; _i++) {
                uniName = _ref2[_i];
                prog.uniforms[uniName] = gl.getUniformLocation(prog, uniName);
              }
              _ref3 = shader.atts;
              _results2 = [];
              for (_j = 0, _len2 = _ref3.length; _j < _len2; _j++) {
                attName = _ref3[_j];
                _results2.push(gl.enableVertexAttribArray(prog.atts[attName] = gl.getAttribLocation(prog, attName)));
              }
              return _results2;
            }
          }).call(this));
        }
        return _results;
      }
    };
    Engine.prototype.getSightDistance = function(eyeHeight) {
      return Math.sqrt((2 * smio.gfx.UniverseSceneNode.consts.earthRadius * eyeHeight) + (eyeHeight * eyeHeight));
    };
    Engine.prototype.isKeyPressed = function(keyCode) {
      if (!(keyCode != null)) {
        return this.pressedKeys.length;
      } else {
        return _.contains(this.pressedKeys, keyCode);
      }
    };
    Engine.prototype.handleKeyDown = function(e) {
      var c;
      if (!_.contains(this.pressedKeys, e.keyCode)) {
        this.pressedKeys.push(e.keyCode);
      }
      if ((c = String.fromCharCode(e.keyCode)) && (c = c.toUpperCase())) {
        if (c === 'C') {
          this.universe.camFar = !this.universe.camFar;
        }
      }
      return Engine.__super__.handleKeyDown.call(this, e);
    };
    Engine.prototype.handleKeyUp = function(e) {
      if (_.contains(this.pressedKeys, e.keyCode)) {
        this.pressedKeys = _.without(this.pressedKeys, e.keyCode);
      }
      return Engine.__super__.handleKeyUp.call(this, e);
    };
    Engine.prototype.handleMouseMove = function(e) {
      this.universe.mouseLook = true;
      return Engine.__super__.handleMouseMove.call(this, e);
    };
    Engine.prototype.play = function() {
      var getAnimFrame, now, timings;
      getAnimFrame = this.requestAnimFrame;
      if (this.gl.isContextLost()) {
        alert('context lost');
      }
      timings = {
        now: (now = new Date().getTime()),
        last: this.lastDrawTime,
        dif: now - this.lastDrawTime
      };
      this.drawTimes.push(timings.dif);
      this.draw(timings);
      this.lastDrawTime = now;
      return getAnimFrame(this.play);
    };
    Engine.prototype.popMatrix = function() {
      return this.modelViewMatrix = this.matrixStack.pop();
    };
    Engine.prototype.pushMatrix = function() {
      var copy;
      copy = mat4.create();
      mat4.set(this.modelViewMatrix, copy);
      return this.matrixStack.push(copy);
    };
    Engine.prototype.updateCanvasSize = function() {
      var height, width, _ref;
      _ref = [this.canvas.width(), this.canvas.height()], width = _ref[0], height = _ref[1];
      this.canvasSize = {
        wpx: this.gl.drawingBufferWidth || this.gl.canvas.width,
        hpx: this.gl.drawingBufferHeight || this.gl.canvas.height,
        w: width,
        h: height,
        w2: width / 2,
        w4: width / 4,
        h22: height / 2.2,
        h15: height / 1.5
      };
      document.title = "" + this.canvasSize.wpx + " x " + this.canvasSize.hpx;
      this.gl.viewport(0, 0, this.canvasSize.wpx, this.canvasSize.hpx);
      return mat4.perspective(45, this.canvasSize.wpx / this.canvasSize.hpx, 0.1, 100.0, this.projectionMatrix);
    };
    return Engine;
  })();
}).call(this);

File: _cscript/client/gfx/Engine.cs
smio = global.smoothio

class smio.gfx.Engine #extends CL3D.CopperLicht

	constructor: (@ctl, cid) ->
		@drawTimes = []
		@lastDrawTime = 0
		@pressedKeys = []
		@matrixStack = []
		@shaders = {}
		@meshes = [new smio.gfx.MeshCube(@)] # new smio.gfx.MeshPyramid(@)
		if (@canvas = $("##{cid}")) and @canvas.length and (@canvEl = @canvas[0]) and @initEngine() and @requestAnimFrame
			@texMan = new smio.gfx.TextureManager(@)
			@texMan.load('stones', '/_/file/images/textures/stones.jpg')
			@texMan.load('wood', '/_/file/images/textures/wood.jpg')
			@texMan.load('sky3', '/_/file/images/textures/sky3.jpg')
			@updateCanvasSize()
			@play()
			return
			@addScene(@scene = new CL3D.Scene())
			@scene.setBackgroundColor(CL3D.createColor(255, 0, 0, 48))
			@scene.getRootSceneNode().addChild(@skyBox = new CL3D.SkyBoxSceneNode())
			@skyBox.getMaterial(5).Tex1 = @getTextureManager().getTexture('/_/file/images/textures/stars.jpg', true)
			@skyBox.getMaterial(4).Tex1 = @getTextureManager().getTexture('/_/file/images/textures/skxup.jpg', true)
			@skyBox.getMaterial(0).Tex1 = @getTextureManager().getTexture('/_/file/images/textures/skx1.jpg', true)
			@skyBox.getMaterial(2).Tex1 = @getTextureManager().getTexture('/_/file/images/textures/skx3.jpg', true)
			@skyBox.getMaterial(1).Tex1 = @getTextureManager().getTexture('/_/file/images/textures/skx2.jpg', true)
			@skyBox.getMaterial(3).Tex1 = @getTextureManager().getTexture('/_/file/images/textures/skx0.jpg', true)
			@scene.getRootSceneNode().addChild(@universe = new smio.gfx.UniverseSceneNode(@))
			@scene.setActiveCamera(@universe.cam)
			@universe.camSettings(@canvEl.width / @canvEl.height, CL3D.degToRad(45), @getSightDistance(20), 1)

	createFragmentShader: (src) =>
		@createShader(@gl.FRAGMENT_SHADER, src)

	createVertexShader: (src) =>
		@createShader(@gl.VERTEX_SHADER, src)

	createShader: (type, src) =>
		if gl = @gl
			shader = gl.createShader(type)
			gl.shaderSource(shader, src)
			gl.compileShader(shader)
			if not (gl.getShaderParameter(shader, gl.COMPILE_STATUS))
				alert(gl.getShaderInfoLog(shader))
		shader

	createVertex: (x, y, z, s, t) ->
		v = new CL3D.Vertex3D(true)
		v.Pos.X = x
		v.Pos.Y = y
		v.Pos.Z = z
		v.TCoords.X = s
		v.TCoords.Y = t
		v

	draw: (timings) =>
		if (gl = @gl) and (canvas = gl.canvas)
			gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT)
			mat4.identity(@modelViewMatrix)
			for name, shaderProg of @shaders
				gl.uniformMatrix4fv(shaderProg.uniforms.pMatrix, false, @projectionMatrix)
			for mesh in @meshes
				if not mesh.hidden
					@drawMesh(gl, mesh, timings)

	drawMesh: (gl, mesh, timings) =>
		@pushMatrix()
		mesh.beforeDraw(gl, timings)
		mat4.translate(@modelViewMatrix, [mesh.posX, mesh.posY, mesh.posZ])
		if mesh.rotX
			mat4.rotateX(@modelViewMatrix, mesh.rotX)
		if mesh.rotY
			mat4.rotateY(@modelViewMatrix, mesh.rotY)
		if mesh.rotZ
			mat4.rotateZ(@modelViewMatrix, mesh.rotZ)
		if mesh.vertexBuffer
			gl.bindBuffer(gl.ARRAY_BUFFER, mesh.vertexBuffer)
			for name, shader of @shaders
				gl.vertexAttribPointer(shader.atts.aVertexPosition, mesh.vertices[0].length, gl.FLOAT, false, 0, 0)
		if mesh.colorBuffer
			gl.bindBuffer(gl.ARRAY_BUFFER, mesh.colorBuffer)
			for name, shader of @shaders
				gl.vertexAttribPointer(shader.atts.aVertexColor, mesh.colors[0].length, gl.FLOAT, false, 0, 0)
		if mesh.texCoordsBuffer
			gl.bindBuffer(gl.ARRAY_BUFFER, mesh.texCoordsBuffer)
			for name, shader of @shaders
				gl.vertexAttribPointer(shader.atts.aTexCoord, mesh.texCoords[0].length, gl.FLOAT, false, 0, 0)
			gl.activeTexture(gl.TEXTURE0)
			gl.bindTexture(gl.TEXTURE_2D, @texMan.textures['sky3'])
			for name, shader of @shaders
				gl.uniform1i(shader.uniforms.uSampler, 0)
		if mesh.indexBuffer
			gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, mesh.indexBuffer)
		for name, shaderProg of @shaders
			gl.uniformMatrix4fv(shaderProg.uniforms.mvMatrix, false, @modelViewMatrix)
		mesh.draw(gl, timings)
		@popMatrix()

	initEngine: =>
		[canvas, gl, names] = [@canvEl, null, ['webgl', 'experimental-webgl', 'webkit-3d', 'moz-webgl']]
		@requestAnimFrame = window.requestAnimationFrame or window.webkitRequestAnimationFrame or window.mozRequestAnimationFrame or window.oRequestAnimationFrame or window.msRequestAnimationFrame
		@modelViewMatrix = mat4.create()
		@projectionMatrix = mat4.create()
		for name in names
			try
				gl = canvas.getContext(name, alpha: true, depth: true, stencil: true, antialias: true, premultipliedAlpha: true, preserveDrawingBuffer: false)
			if gl
				break
		if (@gl = gl)
			@initEngineShaders()
			@initEngineBuffers()
			gl.clearColor(0.1, 0.2, 0.3, 1.0)
			gl.enable(gl.DEPTH_TEST)
			try
				gl.enable(gl.TEXTURE_2D)
		gl

	initEngineBuffers: =>
		if (gl = @gl)
			for mesh in @meshes
				mesh.updateBuffers()

	initEngineShaders: =>
		if (gl = @gl)
			for name, shader of smio.gfx.Shaders
				if not shader.disabled
					vertexShader = @createVertexShader(shader.vertex)
					fragShader = @createFragmentShader(shader.fragment)
					@shaders[name] = prog = gl.createProgram()
					gl.attachShader(prog, vertexShader)
					gl.attachShader(prog, fragShader)
					gl.linkProgram(prog)
					if not gl.getProgramParameter(prog, gl.LINK_STATUS)
						alert('Could not link shader program.')
					gl.useProgram(prog)
					if (not shader.atts) or not shader.atts.length
						shader.atts = ['aVertexPosition']
					else if not _.contains(shader.atts, 'aVertexPosition')
						shader.atts.push('aVertexPosition')
					prog.atts = {}
					prog.uniforms =
						pMatrix: gl.getUniformLocation(prog, 'uPMatrix')
						mvMatrix: gl.getUniformLocation(prog, 'uMVMatrix')
					for uniName in shader.uniforms
						prog.uniforms[uniName] = gl.getUniformLocation(prog, uniName)
					for attName in shader.atts
						gl.enableVertexAttribArray(prog.atts[attName] = gl.getAttribLocation(prog, attName))

	getSightDistance: (eyeHeight) =>
		Math.sqrt((2 * smio.gfx.UniverseSceneNode.consts.earthRadius * eyeHeight) + (eyeHeight * eyeHeight))

	isKeyPressed: (keyCode) =>
		if not keyCode? then @pressedKeys.length else _.contains(@pressedKeys, keyCode)

	handleKeyDown: (e) =>
		if not _.contains(@pressedKeys, e.keyCode)
			@pressedKeys.push(e.keyCode)
		if (c = String.fromCharCode(e.keyCode)) and (c = c.toUpperCase())
			if c is 'C'
				@universe.camFar = not @universe.camFar
		super(e)

	handleKeyUp: (e) =>
		if _.contains(@pressedKeys, e.keyCode)
			@pressedKeys = _.without(@pressedKeys, e.keyCode)
		super(e)

	handleMouseMove: (e) =>
		@universe.mouseLook = true
		super(e)

	play: =>
		getAnimFrame = @requestAnimFrame
		if @gl.isContextLost()
			alert 'context lost'
		timings = now: (now = new Date().getTime()), last: @lastDrawTime, dif: (now - @lastDrawTime)
		@drawTimes.push(timings.dif)
		@draw(timings)
		@lastDrawTime = now
		getAnimFrame(@play)

	popMatrix: =>
		@modelViewMatrix = @matrixStack.pop()

	pushMatrix: =>
		copy = mat4.create()
		mat4.set(@modelViewMatrix, copy)
		@matrixStack.push(copy)

	updateCanvasSize: =>
		[width, height] = [@canvas.width(), @canvas.height()]
		@canvasSize = wpx: @gl.drawingBufferWidth or @gl.canvas.width, hpx: @gl.drawingBufferHeight or @gl.canvas.height, w: width, h: height, w2: (width / 2), w4: (width / 4), h22: (height / 2.2), h15: (height / 1.5)
		document.title = "#{@canvasSize.wpx} x #{@canvasSize.hpx}"
		@gl.viewport(0, 0, @canvasSize.wpx, @canvasSize.hpx)
		mat4.perspective(45, @canvasSize.wpx / @canvasSize.hpx, 0.1, 100.0, @projectionMatrix)

